In [31]:
# Import Libraries
import pandas as pd
import numpy as np
import os



In [33]:
# import products csv
df_prods=pd.read_csv(r"C:\Users\Owner\Downloads\4.3_orders_products\products.csv")

Mixed-Type Data

In [36]:
# create a dataframe
df_test = pd.DataFrame()

In [38]:
# create a mixed type column
df_test['mix'] = ['a','b',1,True]

In [40]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [42]:
# checks if the data frame has any mixed type columns\n",
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


C:\Users\Owner\AppData\Local\Temp\ipykernel_24340\2426354186.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)


In [44]:
# changed the data type for the mixed type column - now it will not be mixed,the contents will be string
df_test['mix'] = df_test['mix'].astype('str')

Missing Values

In [47]:
# finds the number of missing values
df_prods.isnull().sum()


product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [49]:
# to view the number of missing values by creating a new dataframe
df_nan = df_prods[df_prods['product_name'].isnull() == True]
 

In [51]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [53]:
# to compare the number of rows in your current dataframe with the number in your subset once the missing rows have been removed
df_prods.shape

(49693, 5)

In [55]:
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [57]:
df_prods_clean.shape

(49677, 5)

Duplicate Values

In [60]:
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [62]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [64]:
df_prods_clean.shape

(49677, 5)

In [66]:
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [68]:
df_prods_clean_no_dups.shape

(49672, 5)

Export

In [78]:
import os
import pandas as pd

# Define the path variable
path = r'C:\Users\\Owner\07-2024 Instacart Basket Analysis\02 Data'

# Assuming df_prods is your DataFrame
df_prods.to_csv(os.path.join(path, 'Prepared Data', 'products_checked.csv'))

Task 4.5

Step 2

In [82]:
# import orders csv
df_ords=pd.read_csv(r"C:\Users\Owner\Downloads\4.3_orders_products\orders.csv")

In [83]:
df_ords.describe()

order_id       user_id  order_number     order_dow  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01  2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01  2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00  0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00  1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01  3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01  5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02  6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

'days_since prior order': Has a count less than all other rows indicating missing values. It also has a measure of zero indicating that some orders could be placed within the same day. The max value is 30, which may indicate that orders are capped at 30 days or 1 month
'order_number': Has a max value of 100, which may mean orders are capped at 100 items
'order_day of the week': Has min value of 0 and max value of 6 meaning the seven numbers (0-6) represent days of the week
'order_hour_of_day': Has a min value of 0 and max value of 23 meaning the twenty-four numbers (0-23) represent 24 hrs of the day.

Step 3

In [92]:
for col in df_ords.columns.tolist():
    weird = (df_ords[col].map(type) != df_ords[col].iloc[0].__class__).any()
    if weird:
        print(col)

order_id
user_id
order_number
order_dow
order_hour_of_day
days_since_prior_order


In [94]:
df_ords.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3421083 entries, 0 to 3421082
Data columns (total 7 columns):
 #   Column                  Dtype  
---  ------                  -----  
 0   order_id                int64  
 1   user_id                 int64  
 2   eval_set                object 
 3   order_number            int64  
 4   order_dow               int64  
 5   order_hour_of_day       int64  
 6   days_since_prior_order  float64
dtypes: float64(1), int64(5), object(1)
memory usage: 182.7+ MB


Step 5 

In [98]:
# Find missing values takes sum of Trues in a  column
df_ords.isnull().sum()

order_id                       0
user_id                        0
eval_set                       0
order_number                   0
order_dow                      0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

The variable that has missing values is the 'days_since_prior_order' which has 206209 values missing.

Step 6 

In [100]:
# Create subset of the dataframe that contains the nulls
df_ords_nan = df_ords[df_ords['days_since_prior_order'].isnull() == True]

In [102]:
df_ords_nan

order_id  user_id eval_set  order_number  order_dow  \
0         2539329        1    prior             1          2   
11        2168274        2    prior             1          2   
26        1374495        3    prior             1          1   
39        3343014        4    prior             1          6   
45        2717275        5    prior             1          3   
...           ...      ...      ...           ...        ...   
3420930    969311   206205    prior             1          4   
3420934   3189322   206206    prior             1          3   
3421002   2166133   206207    prior             1          6   
3421019   2227043   206208    prior             1          1   
3421069   3154581   206209    prior             1          3   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
11                      11                     NaN  
26                      14                     NaN  
39                      11                     NaN  
45                      12                     NaN  
...                    ...                     ...  
3420930                 12                     NaN  
3420934                 18                     NaN  
3421002                 19                     NaN  
3421019                 15                     NaN  
3421069                 11                     NaN  

[206209 rows x 7 columns]

In [108]:
df_ords['days_since_prior_order'].fillna(7, inplace=True)

C:\Users\Owner\AppData\Local\Temp\ipykernel_24340\4035663876.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df_ords['days_since_prior_order'].fillna(7, inplace=True)


In [110]:
df_ords[df_ords['order_number']==1]

order_id  user_id eval_set  order_number  order_dow  \
0         2539329        1    prior             1          2   
11        2168274        2    prior             1          2   
26        1374495        3    prior             1          1   
39        3343014        4    prior             1          6   
45        2717275        5    prior             1          3   
...           ...      ...      ...           ...        ...   
3420930    969311   206205    prior             1          4   
3420934   3189322   206206    prior             1          3   
3421002   2166133   206207    prior             1          6   
3421019   2227043   206208    prior             1          1   
3421069   3154581   206209    prior             1          3   

         order_hour_of_day  days_since_prior_order  
0                        8                     7.0  
11                      11                     7.0  
26                      14                     7.0  
39                      11                     7.0  
45                      12                     7.0  
...                    ...                     ...  
3420930                 12                     7.0  
3420934                 18                     7.0  
3421002                 19                     7.0  
3421019                 15                     7.0  
3421069                 11                     7.0  

[206209 rows x 7 columns]

Step 7 

In [112]:
# Create subset of df_ords that contains only rows of duplicates
df_ords_dups = df_ords[df_ords.duplicated()]

In [114]:
df_ords_dups

Empty DataFrame
Columns: [order_id, user_id, eval_set, order_number, order_dow, order_hour_of_day, days_since_prior_order]
Index: []

No duplicates because it is empty - however, if there were duplicates using df.drop_duplicates() to create a sub data frame would be appropriate

Step 9

In [116]:
import os
import pandas as pd

# Define the path variable
path = r'C:\Users\Owner\07-2024 Instacart Basket Analysis\02 Data'

# Assuming df_ords is your DataFrame
df_ords.to_csv(os.path.join(path, 'Prepared Data', 'orders_cleaned.csv'))

In [118]:
import os
import pandas as pd

# Define the path variable
path = r'C:\Users\Owner\07-2024 Instacart Basket Analysis\02 Data'

# Assuming df_prods is your DataFrame
df_prods.to_csv(os.path.join(path, 'Prepared Data', 'products_rechecked.csv'))